python/ray/worker.py

```diff
@@ -222,6 +223,24 @@ def __init__(self):
         self.make_actor = None
         self.actors = {}
         self.actor_task_counter = 0
+        # This is a dictionary mapping actor ID to handle, one for every active
+        # handle that was created by the currently executing task.  Entries are
+        # automatically removed from the dictionary when the created actor
+        # handle goes out of scope (i.e. when the current task ends). If the
+        # current task "forks" a handle by passing it to a remote task, the
+        # forked handle is "joined" into the original handle in this dictionary
+        # after retrieving the result of the remote task.  The next method
+        # submitted on the original handle will then execute after methods
+        # submitted by the forked handle. This guarantees consistent execution
+        # ordering between callers of the same actor.
+        # NOTE(swang): Execution ordering can only be guaranteed if the remote
+        # task's results are actually retrieved through a ray.get. There are no
+        # guarantees for ray.wait.
+        self.created_actor_handles = weakref.WeakValueDictionary()
```

Comment:
Do we ever remove stuff from this dictionary?

Reply:
Yes, they get removed when the actor handle goes out of scope. I'm a little wary of depending on Python here, but I think this actually makes the most sense.

Reply:
Would it make sense to add a test testing that things are actually going out of scope? Or is that covered by the current tests indirectly in some way?

Reply:
Hmm, I can't think of a good Python test for this without looking into the worker's attributes. Would that work?

Reply:
Maybe something like

```python
ray.init(num_workers=1, num_cpus=1)

@ray.remote
class Foo(object):
    pass

a1 = Foo.remote()
a2 = Foo.remote()

@ray.remote
def f(num_unique_actor_handles, *actor_handles):
    assert len(ray.worker.global_worker.created_actor_handles) == n

ray.get(f.remote(0))
ray.get(f.remote(1, a1))
ray.get(f.remote(0))
ray.get(f.remote(2, a1, a2))
ray.get(f.remote(1, a1, a1))
ray.get(f.remote(0))
```

Reply:
Possibly another test where `f` is an actor method as well.